Viewshed geoprocessing load failure fix

sclaridge opened the pull request:
PR to add a progress bar to Viewshed geoprocessing sample to disable map interaction when the `GeoprocessingJob` is loading. 

Previously, an error is thrown when the map is tapped while the job is loading. The progress bar is similar to the one used in the [Browse WFS layers](https://github.com/Esri/arcgis-runtime-samples-android/blob/main/java/browse-wfs-layers/src/main/res/layout/activity_main.xml) sample.

shubham7109 commented on java/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.java:
```diff
@@ -73,6 +75,7 @@ protected void onCreate(Bundle savedInstanceState) {
     // location services
     ArcGISRuntimeEnvironment.setApiKey(BuildConfig.API_KEY);
 
+    mLoadingView = findViewById(R.id.loadingView);
```
```suggestion
    // inflate MapView and the LoadingView from layout
    mMapView = findViewById(R.id.mapView);
    mLoadingView = findViewById(R.id.loadingView);
```
Initialize the LoadingView and MapView just to keep code easier to read

TADraeseke commented on java/viewshed-geoprocessing/src/main/res/layout/activity_main.xml:
```diff
@@ -12,4 +12,30 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:background="@color/loading_view_bg"
+        android:clickable="true"
+        android:focusable="true"
```
these may not be necessary

sclaridge replied:
@TADraeseke I think these are required to disable map interactions. When removed, the geoprocessing job fails when the map is tapped while the job is loading.

TADraeseke commented on java/viewshed-geoprocessing/src/main/res/layout/activity_main.xml:
```diff
@@ -12,4 +12,30 @@
         android:layout_height="match_parent">
     </com.esri.arcgisruntime.mapping.view.MapView>
```
Not anything you did, but to clean up here we can just end the MapView block with />

TADraeseke reviewed (changes requested):
Just spotted a few small things. Thanks @sclaridge!

sclaridge commented on java/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.java:
```diff
@@ -189,6 +194,7 @@ private void performGeoprocessing(final FeatureCollectionTable featureCollection
         // listen for job success
         mGeoprocessingJob.addJobDoneListener(new Runnable() {
           @Override public void run() {
+            mLoadingView.setVisibility(View.GONE);
```
Added comments to both Java and Kotlin samples.

sclaridge commented:
@shubham7109 @TADraeseke I've gone ahead and addressed your comments. I've also added the `loadingView` to the Kotlin sample. Let me know what you think!

TADraeseke commented on java/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.java:
```diff
@@ -55,176 +57,185 @@
+        ArcGISMap map = new ArcGISMap(BasemapStyle.ARCGIS_TOPOGRAPHIC);
+        // set the map to be displayed in this view
+        mMapView.setMap(map);
+        mMapView.setViewpoint(new Viewpoint(45.3790902612337, 6.84905317262762, 100000));
```
```suggestion
        mMapView.setViewpoint(new Viewpoint(45.3790, 6.8490, 100000));
```

Point doesn't need to be so precise

sclaridge replied:
I've changed this in both samples.

TADraeseke commented on java/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.java:
```diff
@@ -55,176 +57,185 @@
+                mGeoprocessingJob.start();
+
+                // listen for job success
+                mGeoprocessingJob.addJobDoneListener(new Runnable() {
+                    @Override
+                    public void run() {
```
this can all be a () -> lambda

TADraeseke commented on java/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.java:
```diff
@@ -55,176 +57,185 @@
+                    }
+                });
+            } catch (InterruptedException | ExecutionException e) {
+                e.printStackTrace();
```
Look at other samples for how we handle errors. We write them as `Toast` and out to the Logcat

TADraeseke replied:
https://devtopia.esri.com/runtime/common-samples/wiki/Android-Samples-Repo%3A-A-Contribution-Guide#error-handling

TADraeseke commented on java/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.java:
```diff
@@ -55,176 +57,185 @@
+        });
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        mMapView.pause();
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        mMapView.resume();
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        mMapView.dispose();
+    }
```
The order here should match the life cycles here

https://devtopia.esri.com/runtime/common-samples/wiki/Android-Samples-Repo%3A-A-Contribution-Guide#error-handling

TADraeseke reviewed (changes requested):
Thanks @sclaridge just some small stuff and a few links to the contribution guide.

Overall the changes look good!

sclaridge commented on kotlin/viewshed-geoprocessing/src/main/res/layout/activity_main.xml:
```diff
@@ -1,4 +1,45 @@
-<com.esri.arcgisruntime.mapping.view.MapView xmlns:android="http://schemas.android.com/apk/res/android"
-    android:id="@+id/mapView"
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.coordinatorlayout.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
ConstraintLayout works well!

sclaridge commented:
@TADraeseke Thanks for the Contribution guide links! I've fixed the errors and addressed your comments - ready for a re-review!

shubham7109 commented on kotlin/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.kt:
```diff
@@ -191,6 +199,9 @@ class MainActivity : AppCompatActivity() {
                 // listen for job success
                 geoprocessingJob?.addJobDoneListener {
+                    // hide the LoadingView when job is done loading
+                    loadingView!!.visibility = View.GONE
```
this ^

TADraeseke reviewed (approved):
LGTM! Thanks @sclaridge.

I'll move the issue into to verify and tag @saratchandrakarumuri for his verification.

saratchandrakarumuri reviewed (approved):
Both java and kotlin versions are working as expected
